Initial Code Review

hepcat72 opened the pull request:
The changes here represent everything that has been done to create this package thus far.  I created an empty branch off the initial commit called audit and this PR is a merge of master's HEAD into the audit branch.  The main thing to review is:

- `htseq2multx.pl`

It would also be helpful to review:

- `tests/run_tests.sh`

Anything else is a bonus.

hepcat72 committed:
Filled in basic description.

Files checked in: README.md

hepcat72 committed:
Initial code and tests checkin.

Everything is currently working.

Files checked in: .gitignore htseq2multx.pl tests/run_tests.sh tests/test_data/*

hepcat72 committed:
ds store.

Files checked in: .gitignore

hepcat72 committed:
Tidied up the code a bit.

Files checked in: htseq2multx.pl

hepcat72 committed:
Added comments and re-organized subs.

Files checked in: .gitignore htseq2multx.pl

hepcat72 committed:
Removed test outputs.

lparsons commented on README.md:
```diff
@@ -1,2 +1,12 @@
+Part of the supported functionality is the demultiplexing of pooled genomic samples.  The initial implementation of this functionality utilized an in-house python tool called [barcode_splitter](https://bitbucket.org/princeton_genomics/barcode_splitter/) written in the interpreted scripting language called Python.  The performance of barcode_splitter appears to be I/O bound, and as sample capacity ...
+
+fastq-multx is a demultiplexing tool written in C and performs several orders of magnitude faster.  Since HTSEQ is highly integrated with barcode_splitter, creating a wrapper to make fastq-multx look and act like barcode_splitter requires much less effort than altering HTSEQ to use fastq-multx.
+
```
Might be nice to have some simple usage examples and a note that this requires `fastq_multx` to be on the path (I assume).

Also, would be good to have a LICENSE file.

hepcat72 replied:
Sorry.  I should have marked this PR as a WIP and I should have realized that my `I'm not done with the repo` portion of the commentary accompanying the PR was not enough.

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+my $debug           = 0;
+my $check_deflines  = 0;
+my $tmp_files       = [];
+my($bcfile,$bcfilefqmx,$stodout,$prefix,$suffix,$format,$gzipin,$deflinesep);
```
`$stodout` is declared but not used, probably meant `$stdout`

hepcat72 replied:
Whoops.  Good catch.

- [x] Remove `$stodout`

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    #Handle the version option
+    if($version == 1)
+      {print("$BCSVERSION\n")}
+    elsif($version)
+      {print("barcode_splitter, version $BCSVERSION\n",
+             "htseq2multx, version $VERSION\n",
+             "fastq-multx, version $FQMXVERSION\n")}
```
These seem to print versions that are hardcoded, not derived from the executables being run.

hepcat72 replied:
Huh.  Yeah, that's a good point.  I didn't think of that.  The `BCSVERSION` was intended to be hard-coded because that's the version this emulates, but I agree that `FQMXVERSION ` should reflect the version used.

- [x] Make FQMXVERSION into the fastq-multx minimum required version
- [x] Add a method to retrieve and spit back the fastq-multx version used

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+                      'split_all'       => \$split_all,
+                      'format=s'        => \$format,
+                      'gzipin'          => \$gzipin,
+                      '<>'              => sub {push(@$fastq_files,$_[0])},
```
Minor, but my general preference is for (for clarity of deref precedence):

```suggestion
                      '<>'              => sub {push(@{ $fastq_files },$_[0])},
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+                      'check-deflines'  => \$check_deflines
+                     };
+
+    if(!GetOptions(%$OptionHash))
```
Again, minor preference issue:
```suggestion
    if(!GetOptions(%{ $OptionHash }))
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    ## Ensure basically valid values for required options are supplied
+    ##
+
+    if(scalar(@$fastq_files) == 0)
```
```suggestion
    if(scalar(@{ $fastq_files }) == 0)
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+        exit(1);
+      }
+
+    if(scalar(@$idxread) == 0)
```
```suggestion
    if(scalar(@{ $idxread }) == 0)
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+        exit(1);
+      }
+
+    if(scalar(grep {$_ < 1 || $_ > scalar(@$fastq_files)} @$idxread))
```
```suggestion
    if(scalar(grep {$_ < 1 || $_ > scalar(@{ $fastq_files })} @{ $idxread }))
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+      }
+
+    #Initialize the data structures/hashes used, e.g. index_hash, bchash,
+    #samplehash, and read in & convert eh barcode file
```
```suggestion
    #samplehash, and read in & convert the barcode file
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    #Initialize the data structures/hashes used, e.g. index_hash, bchash,
+    #samplehash, and read in & convert eh barcode file
+    my $order = 1;
+    foreach my $fqf_num (@$idxread)
```
```suggestion
    foreach my $fqf_num (@{ $idxread })
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    elsif(fqmxVersionInsufficient($multx_version,$supported_version))
+      {
+        help2();
+        if(scalar(@$multx_version) < 1)
```
```suggestion
        if(scalar(@{ $multx_version }) < 1)
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+          {print STDERR ('Unable to determine fastq-multx version of ',
+                         "executable [$fastq_multx].  Unable to proceed.\n")}
+        else
+          {print STDERR ('Version [',join('.',@$multx_version),
```
```suggestion
          {print STDERR ('Version [',join('.',@{ $multx_version) },
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    return($insufficient);
+  }
+
+sub bcs2fqmxBCFile
```
Would be helpful to show a simple example of the input/output of this function. I realize it's converting the barcode file format, but it's kinda hard to see what's its doing at a glance.

hepcat72 replied:
- [x] Add documentation for bcs2fqmxBCFile

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+          }
+      }
+
+    my $missing_infiles = [grep {!-e $_} @$fastq_files];
```
```suggestion
    my $missing_infiles = [grep {!-e} @{ $fastq_files }];
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+      }
+
+    my $missing_infiles = [grep {!-e $_} @$fastq_files];
+    if(scalar(@$missing_infiles))
```
```suggestion
    if(scalar(@{ $missing_infiles }))
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    my $missing_infiles = [grep {!-e $_} @$fastq_files];
+    if(scalar(@$missing_infiles))
+      {
+        print STDERR ('ERROR: ',scalar(@$missing_infiles),' input files do ',
```
```suggestion
        print STDERR ('ERROR: ',scalar(@{ $missing_infiles }),' input files do ',
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    if(scalar(@$missing_infiles))
+      {
+        print STDERR ('ERROR: ',scalar(@$missing_infiles),' input files do ',
+                      'not exist: ',join(',',@$missing_infiles),"\n");
```
```suggestion
                      'not exist: ',join(',',@{ $missing_infiles }),"\n");
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    ## Check the output files
+    ##
+
+    if(!$split_all && scalar(@$fastq_files) == scalar(@$idxread))
```
```suggestion
    if(!$split_all && scalar(@{ $fastq_files }) == scalar(@{ $idxread }))
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+  --version             show program's version number and exit
+
+Barcodes:
+  --bcfile FILE         REQUIRED: Tab delimited file: "Sample_ID <tab>
+                        Barcode_Sequence" Multiple barcode columns with
+                        different barcode lengths allowed, but all barcodes in
+                        each inidividual column must be the same length.
```
I thought a big advantage of `fastq_multx` was to allow barcodes of varying length?

hepcat72 replied:
Right.  I was trying to emulate barcode splitter without making it look any different, so this was copied directly from barcode splitter, though I do keep the ability to handle different length barcodes.  In fact, I "skipped" the same-length barcode error test.  I think you're right about this suggestion.  I will alter the help message in this instance.

- [x] Edit the --bcfile help description

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    my $short_only = $_[0] || 0;
+
+    my $short_msg  = << 'END_SHORT';
+usage: barcode_splitter [-h] [--version] [--bcfile FILE]
```
```suggestion
usage: htseq2multx.pl [-h] [--version] [--bcfile FILE]
```

hepcat72 replied:
As above, perhaps I am taking the emulation too far.  I wonder though if htseq ever parses this line?  @jcmatese ?

- [x] Reconsider usage example

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+      {$split_all = 1}
+
+    if(outfilesExist($bchash,$fastq_files,$split_all))
+      {exit(7)}
```
Consider making some Readonly vars for various exit codes instead of hardcoding the exit code here. *e.g.*
```perl
Readonly::Scalar    $ERROR_OUTFILES_EXIST => 7;
```

```suggestion
      {exit($ERROR_OUTFILES_EXIST)}
```

hepcat72 replied:
I was rather lax in emulating the exit codes in general, but in the cases where I do keep barcode splitter's exit codes, I think this is a good idea.

- [x] Initialize globals for specific codes.

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+  }
+
+#This detects and converts files input via process substitution, and writes them
+#out to temporary files.  Assumes $gzip is true.
```
Slightly confused, I thought it didn't handle gzip input from a pipe (from discussions on Slack, which may be out of date).

hepcat72 replied:
fastq-multx does not, as you recalled correctly, handle this.  That's why it is handled here.  Do you have a suggested wording that would improve this comment?

hepcat72 replied:
- [x] Improve wording of process substitution handling comment

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+  {
+    my $command = $_[0];
+
+    my $producer = new IO::Pipe::Producer();
```
Don't use indirect syntax: http://modernperlbooks.com/mt/2009/08/the-problems-with-indirect-object-notation.html

```suggestion
    my $producer = IO::Pipe::Producer()->new;
```

hepcat72 replied:
Interesting.  I should probably update my module on CPAN too.

- [x] Change `new ...Producer` to `Producer->new`
- [ ] Change the CPAN module documentation to reflect the above

I had noted some other doc improvements that I need to make to that module.  Incidentally, that's the first module I ever put on CPAN.

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    my $producer = new IO::Pipe::Producer();
+    my($stdout_handle,$stderr_handle) = $producer->getSystemProducer($command);
+
+    my $sel = new IO::Select;
```
```suggestion
    my $sel = IO::Select->new;
```

hepcat72 replied:
- [x] Change `new IO::Select` to `new IO::Select->new`

... I know I can commit the suggestion.  I just want to stick with my parens convention.

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+          }
+      }
+
+    if($? || $make_error_fatal)
```
```suggestion
    if($CHILD_ERROR || $make_error_fatal)
```

hepcat72 replied:
- [x] Use english (`$?` -> `$CHILD_ERROR`) (everywhere)

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    if($? || $make_error_fatal)
+      {
+        print STDERR ("ERROR: fastq-multx command failed",
+                      ($? ? " with a non-zero exit code [$?]" .
```
```suggestion
                      ($CHILD_ERROR ? " with a non-zero exit code [$CHILD_ERROR]" .
```

hepcat72 replied:
Noted.

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+      {
+        print STDERR ("ERROR: fastq-multx command failed",
+                      ($? ? " with a non-zero exit code [$?]" .
+                       ($! =~ /./ ? " and error: $!" : '.') : ':'),
```
```suggestion
                       ($ERRNO =~ /./ ? " and error: $ERRNO" : '.') : ':'),
```

hepcat72 replied:
Noted.

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+          {exit(1)}
+      }
+
+    print(fqmx2bcsStdout($stdout,scalar(@$idxread)));
```
```suggestion
    print(fqmx2bcsStdout($stdout,scalar(@{ $idxread })));
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+  {
+    my $fastq_files = $_[0];
+    my $prefix      = $_[1];
+    return() if(scalar(grep {$_ !~ /\.gz$/i} @$fastq_files) == 0);
```
Don't need the `$_` here:

```suggestion
    return() if(scalar(grep {/\.gz$/i} @{ $fastq_files }) == 0);
```

hepcat72 replied:
Removing `$_ !~` makes the filter do the opposite of how it was written.  I know you're trying to eliminate the cryptic `$_` variables, but in this instance, I do not know how to do that without adding cryptic stiff in or after the regex.

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+
+    print OUT (<IN>);
+
+    close(OUT);
+    close(IN);
```
Return value of `close` not checked, consider [`use autodie;`](https://metacpan.org/pod/autodie) for a quick fix.

hepcat72 replied:
- [x] Handle return value of close

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+      {
+        my $template = getOutfileTemplate($fqf_index);
+        next if($template =~ m%n/a%i);
+        foreach my $sample (keys(%$hash),'unmatched')
```
```suggestion
        foreach my $sample (keys(%{ $hash }),'unmatched')
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    my $is_fqmx = 1;
+
+    my $output  = `($fastq_multx 3>&1 1>&2- 2>&3- ) | head -n 2`;
+    if($?)
```
```suggestion
    if($CHILD_ERROR)
```

hepcat72 replied:
Noted.

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    my $output  = `($fastq_multx 3>&1 1>&2- 2>&3- ) | head -n 2`;
+    if($?)
+      {
+        $error = "fastq-multx failed with a non-zero exit code [$?]" .
```
```suggestion
        $error = "fastq-multx failed with a non-zero exit code [$CHILD_ERROR]" .
```

hepcat72 replied:
Noted.

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    if($?)
+      {
+        $error = "fastq-multx failed with a non-zero exit code [$?]" .
+          ($! =~ /./ ? " and error: $!" : '.') . "\n";
```
```suggestion
          ($ERRNO =~ /./ ? " and error: $ERRNO" : '.') . "\n";
```

hepcat72 replied:
Noted.

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    elsif($output =~ /version\D*([\d\.]+)/i)
+      {
+        my $version_string = $1;
+        push(@$version,split(/\./,$version_string));
```
```suggestion
        push(@{ $version,split }(/\./,$version_string));
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    my $reqd = $_[1];
+
+    my $lim =
+      (scalar(@$used) < scalar(@$reqd) ? scalar(@$used) : scalar(@$reqd)) - 1;
```
```suggestion
      (scalar(@{ $used }) < scalar(@{ $reqd }) ? scalar(@{ $used }) : scalar(@{ $reqd })) - 1;
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+
+    #If the required version matches, but the version number of the requirement
+    #is longer than the numbers that were checked
+    if(!$insufficient && $lim == $num_equal && scalar(@$reqd) > $num_equal)
```
```suggestion
    if(!$insufficient && $lim == $num_equal && scalar(@{ $reqd }) > $num_equal)
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    my $dupes       = 0;
+
+    unless(open(BCSBCF,$bcs_file))
+      {print STDERR ("ERROR: Unable to open barcode file: $!") && exit(1)}
```
```suggestion
      {print STDERR ("ERROR: Unable to open barcode file: $ERRNO") && exit(1)}
```

hepcat72 replied:
Noted.

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+      {print STDERR ("ERROR: Unable to open barcode file: $!") && exit(1)}
+
+    unless(open(FQMXBCF,'>',$fqmx_file))
+      {print STDERR ("ERROR: Unable to open fqmx barcode file: $!") && exit(1)}
```
```suggestion
      {print STDERR ("ERROR: Unable to open fqmx barcode file: $ERRNO") && exit(1)}
```

hepcat72 replied:
Noted.

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+        $ln++;
+        chomp;
+        next if(/^\s*#/ || /^\s*$/);
+        my @cols = split(/\t/,$_);
```
Don't need to use `$_`
```suggestion
        my @cols = split(/\t/);
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+                         'but found [',scalar(@cols),"].  Skipping.\n")}
+      }
+
+    close(FQMXBCF);
+    close(BCSBCF);
```
Return value of close not checked (could use autodie, see above).

hepcat72 replied:
Noted.

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+                           map {join(',',@{$uniq_check->{$_}})}
+                           grep {scalar(@{$uniq_check->{$_}}) > 1}
+                           sort {$uniq_check->{$a}->[0] cmp
+                                   $uniq_check->{$b}->[0]} keys(%$uniq_check)),
```
```suggestion
                                   $uniq_check->{$b}->[0]} keys(%{ $uniq_check })),
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+                           sort {$uniq_check->{$a}->[0] cmp
+                                   $uniq_check->{$b}->[0]} keys(%$uniq_check)),
+                      "].\n");
+        exit(12);
```
Consider using a Readonly instead of hardcoded number.

hepcat72 replied:
Noted.

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    if($gzipin)
+      {
+        use IO::Uncompress::Gunzip;
+        my $z = new IO::Uncompress::Gunzip $fastq_files->[0] ||
```
```suggestion
        my $z = IO::Uncompress::Gunzip->new($fastq_files->[0]) ||
```

hepcat72 replied:
Noted.

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+                last;
+              }
+            #Just in case...
+            last if($l > 100);
```
Consider using a Readonly (`$MAX_FASTQ_LINES`).

hepcat72 replied:
- [x] Use global `$MAX_FASTQ_LINES` instead of hard-coded `100`

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+                last;
+              }
+            #Just in case...
+            last if($l > 100);
```
Consider using a Readonly (`$MAX_FASTQ_LINES`).

hepcat72 replied:
Noted.

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+        $output_exists = 1;
+
+        chomp;
+        my @d = split(/\t/,$_);
```
```suggestion
        my @d = split(/\t/);
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+
+        $count_sum += $d[1];
+        my $glob = getGlobPattern($d[0]);
+        push(@$summary_array,[$samplehash->{$d[0]},
```
```suggestion
        push(@{ $summary_array },[$samplehash->{$d[0]},
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+                              $glob]);
+      }
+
+    foreach my $row (@$summary_array)
```
```suggestion
    foreach my $row (@{ $summary_array })
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    my $fatal_pats  = ['Error: number of input files \(\d+\) must match ' .
+                       'number of output files',
+                       'No such file or directory'];
+    my $fatal_pat   = join('|',@$fatal_pats);
```
```suggestion
    my $fatal_pat   = join('|',@{ $fatal_pats });
```

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+    if(!$debug)
+      {
+        my $del_files = [$bcfilefqmx];
+        push(@$del_files,@$tmp_files) if(defined($tmp_files));
```
```suggestion
        if(defined($tmp_files)) { push(@{ $del_files },@{ $tmp_files }) };
```

hepcat72 replied:
- [x] Move post conditionals in front

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+        my $del_files = [$bcfilefqmx];
+        push(@$del_files,@$tmp_files) if(defined($tmp_files));
+        foreach(grep {defined($_) && -e $_} @$del_files)
+          {unlink($_)}
```
```suggestion
          {unlink()}
```

hepcat72 replied:
I tend to use `$_` to be explicit because different methods operate on different defaults in different contexts and it's hard for me to remember.

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,884 @@
+      {
+        my $del_files = [$bcfilefqmx];
+        push(@$del_files,@$tmp_files) if(defined($tmp_files));
+        foreach(grep {defined($_) && -e $_} @$del_files)
```
```suggestion
        foreach(grep {defined() && -e } @{ $del_files })
```

lparsons reviewed (changes requested):
Very thorough job. Most of the comments are minor. I probably shouldn't have included all of the suggestions for each double-sigil deref, they clutter the review, but once I started I thought it best to be consistent.

I wasn't actually able to run the `./run_tests.sh` successfully. Didn't debug too far yet, but here's the output: https://gist.github.com/lparsons/89c46744bf0be957bae7ad501d1c8aa3

hepcat72 commented:
> Very thorough job. Most of the comments are minor. I probably shouldn't have included all of the suggestions for each double-sigil deref, they clutter the review, but once I started I thought it best to be consistent.

I know that my preference for the double-sigil deref is generally in disfavor and that my opinion that they are easier to read (when there's no arrow or index/key) and less cluttery is not a popular one.  In fact, my former boss at Los Alamos, who got me to learn perl, agrees with you.

> I wasn't actually able to run the `./run_tests.sh` successfully. Didn't debug too far yet, but here's the output: https://gist.github.com/lparsons/89c46744bf0be957bae7ad501d1c8aa3

Yeah, since I haven't yet written the Makefile.pl, your run of the tests will fail because:

```
Can't locate IO/Pipe/Producer.pm
```

hepcat72 committed:
Update htseq2multx.pl

Co-authored-by: Lance Parsons <lance@parsonsclan.com>

hepcat72 commented:
Very helpful review @lparsons. Thanks!

hepcat72 commented:
Incidentally @jcmatese, I'll be working on another project today, so reviewing now is fine.  I'll try to remember to change the PR title to prepend `[WIP]` when I start making changes.

hepcat72 commented:
Incidentally, when you go through the review in the github web interface, how do you initiate the independent little code suggestion? Are you like, selecting the lines and using a contextual menu with an edit option or something?  It's pretty nifty you can make those mini-PR's.

lparsons commented:
> Incidentally, when you go through the review in the github web interface, how do you initiate the independent little code suggestion? Are you like, selecting the lines and using a contextual menu with an edit option or something? It's pretty nifty you can make those mini-PR's.

Yeah, it's a very cool feature (and would be useful when not abused as I did). Basically, it works on the line(s) that you are commenting on. Click on the little `+/-` icon and it fills in some markdown automatically. You edit that, and it automatically puts in the little diff. See step number 6 on [Commenting on a pull request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request):

![image](https://user-images.githubusercontent.com/645128/84412825-cafe3180-abdd-11ea-9c5d-7f3fd3a1eef8.png)

lparsons commented:
> Incidentally @jcmatese, I'll be working on another project today, so reviewing now is fine. I'll try to remember to change the PR title to prepend `[WIP]` when I start making changes.

FYI - You can mark PRs as "Draft" as well, see [Introducing Draft Pull Requests](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

Documentation on [Changing the stage of a pull request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request)

![image](https://user-images.githubusercontent.com/645128/84413617-e3bb1700-abde-11ea-93f1-7faf530bc5f9.png)

jcmatese commented:
> > Incidentally @jcmatese, I'll be working on another project today, so reviewing now is fine. I'll try to remember to change the PR title to prepend `[WIP]` when I start making changes.
> 
> FYI - You can mark PRs as "Draft" as well, see [Introducing Draft Pull Requests](https://github.blog/2019-02-14-introducing-draft-pull-requests/).
> 
> Documentation on [Changing the stage of a pull request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request)
> 
> ![image](https://user-images.githubusercontent.com/645128/84413617-e3bb1700-abde-11ea-93f1-7faf530bc5f9.png)

Whoops, I just inadvertently converted this PR into a Draft.  I am not certain I understand the distinction between a draft and a pre-reviewed PR.

jcmatese commented:
Clicking the "ready for review" button, to undo...

hepcat72 commented:
> > Incidentally @jcmatese, I'll be working on another project today, so reviewing now is fine. I'll try to remember to change the PR title to prepend `[WIP]` when I start making changes.
> 
> FYI - You can mark PRs as "Draft" as well, see [Introducing Draft Pull Requests](https://github.blog/2019-02-14-introducing-draft-pull-requests/).
> 
> Documentation on [Changing the stage of a pull request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request)
> 
> ![image](https://user-images.githubusercontent.com/645128/84413617-e3bb1700-abde-11ea-93f1-7faf530bc5f9.png)

Huh.  The convert to draft button didn't seem to appear to do anything, but apparently it worked...  Seems like there should be some sort of confirming feedback.

hepcat72 committed:
First pass at addressing issues from the code review.

Files checked in: .gitignore htseq2multx.pl tests/run_tests.sh

hepcat72 committed:
Install script.

I installed the Readonly module locally to be able to implement that suggestion from the PR.
Associated ignore entries added.

Files checked in: .gitignore htseq2multx.pl Makefile.PL

hepcat72 committed:
Finished addressing PR issues.

- Added Readonly
- Fixed issue #2 (fd error)
- Added install & usage sections to the README
- Added a license
- Added a check on the fastq-multx executable in the Makefile.PL

Files checked in: Makefile.PL README.md htseq2multx.pl LICENSE

hepcat72 commented:
I have not yet created a tag for the fastq-multx release nor updated the documentation for IO::Pipe::Producer, but those are not directly associated with this PR.

Ready for second review from @lparsons and @jcmatese.

lparsons commented on Makefile.PL:
```diff
@@ -0,0 +1,79 @@
+#Makefile.PL version 1.0
+
```
```suggestion

use strict;
use warnings;
```

lparsons commented on Makefile.PL:
```diff
@@ -0,0 +1,79 @@
+
+    if(eval("use File::Which;1;"))
+      {
+	$exe = which($sent_exe);
+	if((!defined($exe) || $exe eq '') && -e $sent_exe && -x $sent_exe)
+	  {$exe = $sent_exe}
+	elsif(!defined($exe))
+	  {$exe = ''}
```
Suggest to convert tabs to spaces throughout the file.
```suggestion
  $exe = which($sent_exe);
  if((!defined($exe) || $exe eq '') && -e $sent_exe && -x $sent_exe)
    {$exe = $sent_exe}
  elsif(!defined($exe))
    {$exe = ''}
```

hepcat72 replied:
Ugh. I must have a mis-config in xemacs.  It's set to not use tabs.  I'll have to figure that out.

- [x] Remove tab characters

hepcat72 replied:
Well, I figured out that [my no-tabs setting is a default that can be changed by various editing modules](https://emacs.stackexchange.com/questions/12173/confusion-by-the-usage-of-indent-tabs-mode-nil) (like the `cperl-mode` module I use, though I'm not sure that's the one that's doing it).  I explored the menus and set something that allowed me to replace the tabs with spaces (they were automatically changed back to tabs before I found that setting).  I'm not entirely sure though that the change I made will be permanent, but I'll revisit if I discover them again.

lparsons commented on Makefile.PL:
```diff
@@ -0,0 +1,79 @@
+	chomp($exe);
+	if($exe =~ /which: Command not found./ || $exe !~ /\S/)
+	  {
+	    print STDERR ("ERROR: System command 'which' does not appear to ",
+                          'exist.  Please install the perl module File::Which.',
+                          "\n");
```
Somehow I ended up getting this error when installing, but I know that `which` is available. At the time, I did not have `fastq-multx` in the path.

hepcat72 replied:
I'll take a look.

- [x] Figure out Lance's `which` error

hepcat72 replied:
I think that the code was based on the behavior of which that no longer seems relevant or applicable to all systems.  I think my fix should resolve it for you.

lparsons commented on Makefile.PL:
```diff
@@ -0,0 +1,79 @@
+   VERSION_FROM    => "$scripts[0]", # finds $VERSION
+   PREREQ_PM       => {'IO::Pipe::Producer'     => 2.01,
+                       'IO::Select'             => 1.21,
+                       'IO::Uncompress::Gunzip' => 2.060,
```
The trailing zero (`0`) requires this be quoted.
```suggestion
                       'IO::Uncompress::Gunzip' => '2.060',
```

hepcat72 replied:
Yeah, I thought it might.  (I don't know why I'm going through these in reverse order, but as below), I was surprised that the check on my system succeeded without the quotes.  I expected it to fail and that I'd have to add quotes, but it didn't.  I will add the quotes.

- [x] Add quotes to the Gunzip module

And since we're requiring gunzip anyway, I realize we can use it in the test, so, I will make your suggested gunzip change below.

lparsons commented on Makefile.PL:
```diff
@@ -0,0 +1,79 @@
+                       'File::Glob'             => 1.17,
+                       'strict'                 => 1.07,
+                       'warnings'               => 1.13,
+                       'English'                => '1.06_01',
```
This seems to be a core module (at least as far back as 5.16.3, which is what we use for HTSEQ). I think this line should be removed.
```
! Installing the dependencies failed: Installed version (1.05) of English is not in range '1.06_01'
```

hepcat72 replied:
I didn't have it.  I had to install English.  I suppose I could reduce the requirement to 1.05.  Frankly, I was surprised/impressed that the check of `'1.06_01'` succeeded for me, given the underscore.

lparsons replied:
That's odd, I don't recall ever installing it (didn't even know you could). At any rate, we'd have to sort out something that works a bit more universally. Not sure what that is right now. Can I ask why you chose that version.as a minimum?

hepcat72 replied:
That's what I tested with.  I don't know.  Maybe I'm not remembering correctly.  Too bad I don't have that terminal open anymore.  Either way, I'm willing to remove it if you'd like.

hepcat72 replied:
I changed it to `1.05`. As long as you're testing with that, then I think that change is appropriate.  Haven't pushed yet.

lparsons replied:
Seems to work fine with `1.05`.

lparsons commented on README.md:
```diff
@@ -1,2 +1,65 @@
+
+### Caveats:
+
+Module dependencies should install automatically, but if any do not and `perl Makefile.PL` issues a warning such as:
+
+    Warning: prerequisite Readonly 2.05 not found.
+
+you can try, for example:
+
+    FTP_PASSIVE=1 PERL_MM_USE_DEFAULT=1 perl -MCPAN -e shell
+    cpan> install Readonly
```
For me, I didn't get any dependencies installed, but this worked fine:
```
cpanm --installdeps --notest .
```

hepcat72 replied:
GTK.  Should I change the README?

lparsons replied:
I think some update is warranted. Not sure how this should work exactly, but what I did worked fine assuming the user has cpanminus installed.

hepcat72 replied:
Right.  I was concerned about that as well.  I'll consult my other packages.  I thought things were supposed to install automatically, but Readonly didn't for me, so perhaps I'm mistaken.  I'm pretty sure I've worked through this before in another distro.  I'll look.

- [x] Figure out automatic module dependency installation

hepcat72 replied:
OK.  Automatic installation is only for the modules included in lib/ in the distro.  The MakeMaker only checks the versions of already installed modules otherwise and reports on missing ones.  Looks like my other distro that handles automatic module installs uses a shell script that installs cpan minus (if not present) and installs dependencies that way.  There are apparently numerous ways to do it.  I'm going to read a bit more and decide which way to go.

hepcat72 replied:
BTW, jury duty starts at 12:30 today.  I might not get this finished today due to that.

lparsons commented on tests/run_tests.sh:
```diff
@@ -0,0 +1,351 @@
+rm -rf "${TEST_OUTPUT}"
+mkdir -p "${TEST_OUTPUT}"
+
+#Perl code to allow fastq-multx debug message to not cause a "difference"
+PERL_DEBUG_FILTER='$c=0;$d="";while(<STDIN>){if($_ !~ /#.*fastq-multx|^1d0$/){print;$c++}elsif(/#.*fastq-multx/){$d.=$_}}if($c && $ARGV[0]){$d =~ s/< #/#/g;print "\033[1;94m$d\033[0m"}exit($c)'
```
Perhaps it would be better to just change the test output file? I'm not sure there is value in keeping the test output files the same when the output is intentionally different.

lparsons replied:
Of course, right after I post the review, I sorted out the issue. `cat` doesn't handle zip files on Ubuntu (not sure it does anywhere). `zcat` isn't always available, at least not as often as `gunzip` is.

hepcat72 replied:
I'm not sure the intent here is properly conveyed.  (And perhaps your second comment here is in the wrong thread?)  This is only for debug mode. I added the ability for the perl script to print the fastq-multx call with a prepended "#" comment character because I wanted to see the call for tests that were failing.  This perl script one-liner is only useful for when `DEBUG` on line 12 is not `0`, which is controlled by the first script argument.  If you wish, I could remove the DEBUG-related stuff entirely.  But I find it handy for figuring out why a test failed.

lparsons replied:
Sorry, guess I missed the debit stuff. That's fine to leave then.

As for the zcat in the test, I realized that the test was intended to test getting gzipped file on input. `zcat` is just `gunzip -c`, so it's unzipping the file. That's not really the intent of the test, but perhaps it just won't work since it relies on filenames to detect gzip input.

hepcat72 replied:
Oh, so you're suggesting that cat relies on the presence of the extension in the same way that fastq-multx does?  That could be.  But even so, that would also apply to barcode_splitter.  Can you successfully run this corresponding test for barcode_splitter on ubuntu?  If the issue on my mac is the same as on ubuntu, then I assume you cannot, since it was `cat` that gave me the error I encountered, which prompted me to use the if conditional to set zcat.  We're talking about test 25, right?

lparsons replied:
No, `cat` doesn't care and works just fine on text or binary files. The original test seems to have been written to test barcode splitter's ability to handle gzipped input from a pipe. Changing it to use `zcar` or `gunzip -c` defeats that purpose.

https://stackoverflow.com/a/8005155

hepcat72 replied:
Oh!  I see!  Right. That didn't occur to me.  So... but like I said, on my mac, cat complains about the compressed file.  Let me rerun and get the error and figure out how to address it correctly.

hepcat72 replied:
So there's no error from cat.  I wasn't remembering correctly.  There's just no gz data output.  I suspect there's more going on here.  I'll dig into it tomorrow.

hepcat72 replied:
Figured it out.  The sub making the temporary file copies wasn't even being called because my conditional was using the opposite logic it should have had - plus, I was deleting the fd file name before the call.  Fix impending...

lparsons reviewed (changes requested):
Looks great overall. I had a few issues installing it, which I've commented on (they all seem like simple fixes).

Test 25 failed for me for some reason. I didn't dig into too far, but if you have trouble replicating or anything let me know. The output of the test is here: https://gist.github.com/lparsons/ce988eb828187499bee9638e08f2f320

Finally, it would be nice to add some CI to run tests (github actions or travis). Not at all required for this PR, so I'll put it as a new issue. Great work! Drop in replacement for barcode splitter that's much faster!

lparsons commented on tests/run_tests.sh:
```diff
@@ -0,0 +1,351 @@
+rm -f ${TEST_OUTPUT}/${TEST}_* 2> /dev/null
+OUTPUTS="BC1-read-1.out.gz BC2-read-1.out.gz BC3-read-1.out.gz BC4-read-1.out.gz unmatched-read-1.out.gz summary.out error.out"
+NOT_OUTPUTS=""
+CAT=cat
+OS=`uname`
+if [ "$OS" == "Darwin" ]; then
+    CAT=zcat
+fi
```
This seems to break for me on Ubuntu (and I imagine it would break on other OSes as well).
```suggestion
CAT="gunzip -c"
```

hepcat72 replied:
Yeah, I added it because it broke on my mac too and I assumed that only macs' version of cat couldn't handle compressed input, because barcode_splitter's test script uses cat on a .gz file and I knew that worked in linux.  I figured it must work on all linux systems... So does Ubuntu have an issue with this test in barcode_splitter's test script?  It must, if this test fails.

hepcat72 replied:
Also, I don't know if assuming users have gunzip would work for everyone either.

hepcat72 committed:
Use gunzip in the test script.

Co-authored-by: Lance Parsons <lance@parsonsclan.com>

hepcat72 committed:
Remove empty line.

Co-authored-by: Lance Parsons <lance@parsonsclan.com>

hepcat72 committed:
Quoted gunzip version.

Co-authored-by: Lance Parsons <lance@parsonsclan.com>

lparsons commented on Makefile.PL:
```diff
@@ -0,0 +1,81 @@
+   NAME            => 'htseq2multx',
+
+   #Add these new keywords supported since 5.005
+   ($[ >= 5.005 ?
```
Not sure how I missed this, but I get this warning:
```
$[ used in numeric ge (>=) (did you mean $] ?) at Makefile.PL line 17.
```

Looks like you meant [`$]` (`$OLD_PERL_VERSION` )](https://perldoc.perl.org/perlvar.html#%24OLD_PERL_VERSION)
```suggestion
   ($] >= 5.005 ?
```

hepcat72 replied:
I did come across this alernate variable the other day.  This code was originally a copy/paste from a long time ago.  I suspected that maybe I should change it, but I haven't dug into it yet.

lparsons replied:
It just seems that `$[` has been deprecated for a while (and perhaps isn't even functional anymore). Even if it was, it seems like it stores the "the index of the first element in an array, and of the first character in a substring".

[perlvar docs for `$[`](https://perldoc.perl.org/perlvar.html#%%5b)

I'm not really sure that changing the variable is really the best thing to do either, it seems from some looking that `use 5.005` would be better (though I wonder if we'd want to bump the version number up? I've only tested in `5.16.3`.

hepcat72 replied:
Yeah.  I'll look into it.  Thanks for the info.

- [x] Look into addressing usage of deprecated `$[`

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,976 @@
+
+sub getFastqMultxCommand
+  {
+    my $command =
+      "$fastq_multx -x -d 1 -B '$bcfilefqmx' -m $mismatches -M $mismatches";
```
```suggestion
    # -x          Don't trim barcodes off before writing out destination
    # -d N        Require a minimum distance of N between the best and next best (2)
    # -B BCFIL    Use barcodes from BCFIL, no determination step, codes in <read1.fq>
    # -m N        Allow N mismatches in union of all indexes, unless -M is supplied. (1)
    # -M M        Allow N,M mismatches in indexes 1,2 respectively (see -m N) (not used)
    my $command =
      "$fastq_multx -x -d 1 -B '$bcfilefqmx' -m $mismatches -M $mismatches";
```

hepcat72 replied:
I like this idea.  I think though that to be accurate to the user's version, maybe I should create a way to output the  executable's usage...? Although perhaps I should output the version that this command was based on...  Hmmm... I'll think about it.

lparsons replied:
There really won't be users of this other than us, I just think it's helpful to document what these options are.

hepcat72 replied:
Right. Well, we could just run fastq-multx to see the usage. But I'll add a comment.

- [x] Comment on the fastq-multx options used

hepcat72 committed:
fastq-multx usage in comments

Co-authored-by: Lance Parsons <lance@parsonsclan.com>

hepcat72 committed:
Wrong variable corrected

Co-authored-by: Lance Parsons <lance@parsonsclan.com>

hepcat72 committed:
Addressed more issues from the PR and fixed a bug with the process substitution code.

Files checked in: .gitignore Makefile.PL README.md htseq2multx.pl tests/run_tests.sh

hepcat72 committed:
Merge branch 'master' of https://github.com/hepcat72/htseq2multx

hepcat72 commented:
I think this is likely the last round.  Great catches @lparsons !

lparsons commented on Makefile.PL:
```diff
@@ -0,0 +1,81 @@
+   ($] >= 5.005 ?
+
+    (AUTHOR        => 'Robert W. Leach (rleach@princeton.edu)',
+     ABSTRACT_FROM => "$scripts[0]") : ()),
```
I get a warning here, not really sure what it means, and perhaps it's not a problem, but thought I'd point it out:
```
WARNING: Setting ABSTRACT via file 'htseq2multx.pl' failed
 at /home/lparsons/.plenv/versions/5.16.3/lib/perl5/5.16.3/ExtUtils/MakeMaker.pm line 609.
```

hepcat72 replied:
Well, now that that `$]` thing is correct, that makes sense. I have to fix this in all my repos.

- [x] Fix `ABSTRACT_FROM`

lparsons commented on Makefile.PL:
```diff
@@ -0,0 +1,81 @@
+   EXE_FILES       => [@scripts]
+  );
+
+my $fastq_multx = getExe($fqmx_name);
+if(!defined($fastq_multx) || $fastq_multx eq '' ||
+   !-e $fastq_multx || !-x $fastq_multx)
+  {print STDERR ("WARNING: The $fqmx_name executable [$fastq_multx] appears ",
+                 'to either not be in your path, not exist, not have execute ',
+                 'permissions, or you have not created a symbolic link named ',
+                 "'$fqmx_name' to the full name of the executable with ",
+                 "version number.  If you have not installed $fqmx_name, ",
+                 "you can find it here: $fqmx_link\n")}
```
I'm still getting a warning here, even though `fastq-multx` is in my PATH.
```
File::Which not found, switching to backup method.
WARNING: The fastq-multx executable [] appears to either not be in your path, not exist, not have execute permissions, or you have not created a symbolic link named 'fastq-multx' to the full name of the executable with version number.  If you have not installed fastq-multx, you can find it here: https://github.com/brwnj/fastq-multx
```

hepcat72 replied:
Resolving simply because this is essentially a duplicate of the issue on 76-77.

lparsons commented on Makefile.PL:
```diff
@@ -0,0 +1,81 @@
+                          "\n");
+            $exe = '';
+          }
+        else
+          {$exe = ''}
```
This doesn't seem quite right. The `which` command was successful, so I think you just want to leave it alone.

hepcat72 replied:
Yeah, that's a mistake.  Honestly, I neglected to test.  Should be an easy fix.

- [x] Fix the make file getExe else

lparsons commented on Makefile.PL:
```diff
@@ -0,0 +1,81 @@
+    $sent_exe    =~ s/ .*//;
+    my $exe      = '';
+
+    if(eval("use File::Which;1;"))
+      {
+        $exe = which($sent_exe);
+        if((!defined($exe) || $exe eq '') && -e $sent_exe && -x $sent_exe)
+          {$exe = $sent_exe}
+        elsif(!defined($exe))
+          {$exe = ''}
+      }
+    else
+      {
+        print STDERR ("File::Which not found, switching to backup method.\n");
+        $exe = `which $sent_exe`;
+        chomp($exe);
+        if($exe =~ /which: Command not found./)
+          {
+            print STDERR ("ERROR: System command 'which' does not appear to ",
+                          'exist.  Please install the perl module File::Which.',
+                          "\n");
+            $exe = '';
+          }
+        else
+          {$exe = ''}
+      }
```
This seems to work for me. Basically, execute the perl or system version of `which` and then test the output.

```suggestion
    if(eval("use File::Which;1;"))
      { $exe = which($sent_exe); }
    else 
      {
        print STDERR ("File::Which not found, switching to backup method.\n");
        $exe = `which $sent_exe`;
        chomp($exe);
        if($exe =~ /which: Command not found./)
        {
          print STDERR ("ERROR: System command 'which' does not appear to ",
                        'exist.  Please install the perl module File::Which.',
                        "\n");
          $exe = '';
        } 
      }
    if((!defined($exe) || $exe eq '') && -e $sent_exe && -x $sent_exe)
      { $exe = $sent_exe }
    elsif(!defined($exe))
      { $exe = '' }
```

hepcat72 replied:
`which` from the module (at least in some versions) empties out valid executables provided with absolute paths if that path is not in PATH, in which case, the recovery I use is necessary.  I have not encountered this issue with system calls to `which`, which is why that check is inside the module conditional.  I will:

- [x] Add a comment to explain the `which` module recovery code

lparsons commented on htseq2multx.pl:
```diff
@@ -0,0 +1,992 @@
+    if(scalar(@$fastq_files) == 0)
+      {
+        help(1);
+        print STDERR ("barcode_splitter: error: the following arguments are ",
```
Should these error messages have `htseq2multx.pl` instead of `barcode_splitter`?

hepcat72 replied:
*sigh*. Well, my whole intent was to make this act and look like barcode_splitter.  I don't know if John is specifically parsing any STDERR.  Probably not, but I was trying to make the output the same.  That was my guiding principle.  I bent that WRT your usage suggestion. So, if you think htseq2multx should intentionally be distinct/different, then I'm fine going with that and I would change things everywhere.  Do you think that would be better?

hepcat72 replied:
I mean, if that's the motivation behind the specific feedback on the 2 issues where you've brought this up, then we should establish this as a change to the requirements:

https://github.com/PrincetonUniversity/htseq/pull/88/files#diff-3c65b2d41a08a66bc1dbaf00ebf795a7R5

Where line 5 states that this was the goal.

hepcat72 replied:
I addressed the latest issues except for this one, for which I am just waiting on clarification.

lparsons replied:
I thought about that. It does make sense to work as much the same as barcode splitter as possible, but I figured since you renamed the executable we weren't going for complete transparency. At any rate, this doesn't seem like a huge issue, I just wasn't sure if you were copy/pasting and forgot to change it, or if you intentionally left this here. I stay keep it as is if you like. What was the other second issue I brought up? (Sorry if I brought this up before and we already discussed).

jcmatese replied:
I don't think any parsing of stderr is occurring (to my knowledge).  I did see an email regarding re-directing stderr to stdout, which _possibly_ might be unwanted.  Previous because was to have a barcode_splitter stdout as a download and important meta-information file, and if unknown errors were being written there, I don't know the consequences.  Perhaps I misinterpreted that email, though.

hepcat72 replied:
I'll keep it for now and create an issue to revisit it. The other issue was the executable name in the usage output, which I changed as per your suggestion.

lparsons reviewed (changes requested):
All the tests pass now! Fantastic. 

The `Makefile.PL` still had some warnings for me. The one about the abstract might not be a big deal, but finding the exe was broken. Pretty small issue though, and I put a suggestion in that worked for me.

Other than that, it looks great!

hepcat72 committed:
Fixed abstract, getExe, and explained conditional.

Files checked in: Makefile.PL

lparsons reviewed (approved):
LGTM! :shipit:

hepcat72 commented:
> LGTM! :shipit:

Thanks for all the attention to detail.  I always learn things from your reviews.

hepcat72 committed:
First attempt at github action to setup perl environment and test.

Files checked in: .github/workflows/setup_perl_env.yml

hepcat72 committed:
Attempt 2

I'm having a hard time following the github action documentation.  I suspect that it assumes a lot of knowledge about how this is all supposed to work.  Having not really done this before, I feel rather lost.  For example, the first error I got after simply copying the yml from the action I wanted and putting it in a yml file, the error I got when it ran was "No event triggers defined in `on`".  There was no `on` clause in the yml from the action I wanted, nor do any of the action documentation items say whether the action yml I get from a distributed action is to go in the same file with the `on` clause I found or where in the hierarchy I should put that yml snippet...  So I'm just trying things.  I wish there was a step by step example for starting from any specific action because I don't want other superfluous actions.  Ugh.

hepcat72 committed:
github action attempt 3

I got a workflow syntax error.  It didn't like the test script call.  I'll try inserting `bash` in front.

hepcat72 committed:
Attempt 4 at github action

I still got a generic yaml syntax error on line 19, so I started to replace the test script call with a simple `echo santity check`, but then I noticed that the interface underlines syntax errors with a red squiggle and when you hover over it, you get an error in a tooltip that said "cannot read an implicit mapping pair; a colon is missed".  Googling that was not directly helpful, but I found a yaml parser online and it was a bit more helpful.  Apparently, I needed to indent the commands 2 more spaces.  So here goes attempt 4...

hepcat72 committed:
github action attempt 5

Apparently I had the perl setup package definition in the wrong place and I needed a `runs-on` defined under the `build` section.

hepcat72 committed:
github actions attempt 6

I got the error "No runner matching the specified labels was found: ubuntu-latest, macos-latest".  I could not find a resolution to this in the documentation or via google, but reading between the lines, I'm guessing that maybe specifying multiple OS's only works for self-hosted actions, so I'm going to try just testing with ubuntu.

hepcat72 committed:
Github action Attempt 6

The last attempt actually ran, but the tests failed and I suspect it's (only?) because fastq-multx was not installed, so this is an attempt to install it from a github repo.  I expect this will fail because I need the g++ compiler, but one step at a time.

hepcat72 committed:
Attempt 8 github action

This doesn't really change anything significant other than removing the 2 different names...

hepcat72 committed:
attempt 9 github actions

Still get the error "every step must define a `uses` or `run` key".  Do I need separate steps?  How do you separate them?  OK, apparently each "-" is a step.  What constitutes a step?  uses? name?...  Maybe if I start each step with a name...?

hepcat72 committed:
attempt 10 github action

Next error: "a step cannot have both the `uses` and `run` keys".  OK.  I cut back the dashes too far.  Let's try using dashes for the run items...

hepcat72 committed:
github action attempt 11

Now I get "The workflow is not valid. .github/workflows/setup_perl_env.yml (Line: 18, Col: 7): 'uses' is already defined".  Sheesh.  OK.  I need another dash for the 2 uses steps.  The "checkout" "uses" is in there twice anyway... maybe I don't even need it the second time even though it's not a repo this time...  Let's try taking that out.

hepcat72 committed:
attempt 12

Hmmm...  It seemed to get farther this time...  I clicked on the build and this error was in the history:

1 Run make
5 make: *** No targets specified and no makefile found.  Stop.
6 ##[error]Process completed with exit code 2.

Maybe I shouldn't have taken out the cd.

hepcat72 committed:
github action attempt 13

Next error:

1 Run cpanm --installdeps --notest .
6 --> Working on .
7 Configuring /home/runner/work/htseq2multx/htseq2multx ... N/A
8 ! Configuring . failed. See /home/runner/.cpanm/work/1593469259.2782/build.log for details.
9 ##[error]Process completed with exit code 1.

Huh, cpanm had worked before I started trying to add fastq-multx...  Maybe that second checkout was in fact needed.

hepcat72 committed:
github action attempt 14

OK great.  It's running, but it still cannot find fastq-multx.  Maybe since the path added is relative, and I cd into tests, I need a .. in front in the path.

hepcat72 committed:
github action attempt 15

Not sure why it's running the tests twice (I'm going to ignore that for now), but also, it's still not finding the fastq-multx executable.  So I tried moving the path manipulation to the perl area.

hepcat72 committed:
github action attempt 16

Still not finding the fastq-multx executable.  According to:

https://github.community/t/adding-a-binary-to-the-path-from-a-docker-github-action-for-use-by-later-workflow-steps/16922/2

I need to use $GITHUB_WORKSPACE.

hepcat72 committed:
gh action attempt 17

Still not getting the executable.  Looking at the example again, I see I didn't have things int the same order.  Doesn't seem like it should make a difference, but here goes...

hepcat72 committed:
gha attempt 18

Sigh.  Still not finding it.  Rearrange some more?  I'm running out of ideas and google is no help...

lparsons committed:
Fixup github action to install and run tests

Currently, test_5 fails

hepcat72 committed:
Tried removing clean

Clean may not be necessary.
I also edited step names.

hepcat72 committed:
Adding debug arg to tests

Going to see if the debug output can explain why tests 5 & 25 fail.  It looks like those are the only tests where line 42 (containing gzip in a process substitution call) executes, so I figure that must have something to do with it.

hepcat72 committed:
Debug level 3

This adds set -x.
Fixed a name typo.

hepcat72 committed:
Seeing if commenting this code fixes tests 5 & 25

These tests end up generating a broken stdout pipe error from gzip, which must be coming from fastq-multx.  The tests only fail in the github action, so I'm committing this commenting of those lines to test to see if it resolves the issue.

hepcat72 committed:
Removing debug arg

hepcat72 committed:
Test omission of gzip broken pipe error

The error doesn't seem to have a deleterious effect and only occurs on certain systems.

hepcat72 committed:
Work-around for gzip stdout broken pipe error on some systems

hepcat72 committed:
Attempt tp properly solve an issue where fastq-multx's gzip system call complains about a broken stdout pipe in the github action.

Files checked in: htseq2multx.pl

hepcat72 committed:
Re-removed the close of stdout/err of the system call.

Files checked in: htseq2multx.pl

hepcat72 committed:
Trying latest perl version

This is to see if it fixes the fastq-multx gzip issue.

hepcat72 committed:
Reverted perl version

hepcat72 committed:
Added debug arg to obtain fastq-multx command of failed test

hepcat72 committed:
Added test of fastq-multx alone to rule out it as the source of the github action that issues the gzip broken pipe error.

Files checked in: tests/run_tests.sh tests/test_data/barcode_splitter_barcodes_fqmx.txt tests/test_data/test_26_BC1-read-1.out.gz tests/test_data/test_26_BC2-read-1.out.gz tests/test_data/test_26_BC3-read-1.out.gz tests/test_data/test_26_BC4-read-1.out.gz tests/test_data/test_26_error.out tests/test_data/test_26_summary.out tests/test_data/test_26_unmatched-read-1.out.gz

hepcat72 committed:
Merge branch 'perlaction' of https://github.com/hepcat72/htseq2multx into perlaction

hepcat72 committed:
Removed debug arg

hepcat72 committed:
Another attempt to fix the gzip broken stdout pipe error in github actions.

Files checked in: htseq2multx.pl

hepcat72 committed:
Merge branch 'perlaction' of https://github.com/hepcat72/htseq2multx into perlaction

hepcat72 committed:
Ruled out the script as the source of the gzip broken pipe error from fastq-multx.

Test 26 proved that the error is isolated to fastq-multx. I commented out the test because the behavior is inconsistent on different systems. I also reverted code I'd added to work-around and debug this issue.

Files checked in: htseq2multx.pl tests/run_tests.sh

hepcat72 committed:
Merge pull request #6 from hepcat72/perlaction

Combination of perlaction and perlaction_lp branches to add continuous integration.

hepcat72 commented:
*nudge* @jcmatese

jcmatese reviewed (approved):
seems comprehensive, and nothing jumped  out at me that required changing prior testing real data/use-cases in development

hepcat72 commented:
Thanks @jcmatese!
